### PR TITLE
Correctly handle return values

### LINKED ISSUE #6: Rework error handling
it seems that every function in roc has it own intricate error return. Just one-for-all func `convertToErr` does not seem to suffice. We need to thoroughly go through every function and emulated the interface.

### PR DESCRIPTION
**Pull request by Asalle**
I adapted the return errors of roc-go to the actual returns from c functions. Please take a look at the test, it somehow fails. Could you check if I'm passing the address conrrectly?

Fixes #6.

### CODE REVIEW

**Comment by gavv** [comment]
> Please take a look at the test, it somehow fails. Could you check if I'm passing the address conrrectly?

That's interesting! I've added debug logs:

```
diff --git a/src/lib/src/address.cpp b/src/lib/src/address.cpp
index 4bebe24e..637a943f 100644
--- a/src/lib/src/address.cpp
+++ b/src/lib/src/address.cpp
@@ -34,28 +34,36 @@ address::SocketAddr& get_address(roc_address* address) {
 }
 
 int roc_address_init(roc_address* address, roc_family family, const char* ip, int port) {
+    roc_log(LogError, "woooooooooooo");
     if (sizeof(roc_address) < sizeof(address::SocketAddr)) {
+        roc_log(LogError, "woooooooooooo1");
         return -1;
     }
 
     if (!address) {
+        roc_log(LogError, "woooooooooooo2");
         return -1;
     }
 
     if (!ip) {
+        roc_log(LogError, "woooooooooooo3");
         return -1;
     }
 
     if (port < 0 || port > USHRT_MAX) {
+        roc_log(LogError, "woooooooooooo4");
         return -1;
     }
 
     address::SocketAddr& sa = *new (address_payload(address)) address::SocketAddr;
 
     if (family == ROC_AF_AUTO || family == ROC_AF_IPv4) {
+        roc_log(LogError, "aaa --%s-- --%d--", ip, port);
         if (sa.set_host_port_ipv4(ip, port)) {
+            roc_log(LogError, "bbb");
             return 0;
         }
+        roc_log(LogError, "ccc");
     }
 
     if (family == ROC_AF_AUTO || family == ROC_AF_IPv6) {
@@ -64,6 +72,8 @@ int roc_address_init(roc_address* address, roc_family family, const char* ip, in
         }
     }
 
+    roc_log(LogError, "woooooooooooo5");
+
     return -1;
 }
```

And run the test:

```
go test -v .          
=== RUN   Test_NewAddress
=== RUN   Test_NewAddress/correct_ip_is_parsed_with_no_error
12:08:54.679 [err] roc_lib: woooooooooooo
12:08:54.679 [err] roc_lib: aaa --127.0.0.1244140625: inuse: : status=Bassa_VahBhaiksukiBigEndianCuneiformDiacriticGC (idle)Hex_DigitInheritedInstMatchInstRune1InterfaceKhudawadiLINUX_2.6MalayalamMongolianNabataeanPalmyreneParseBoolParseUintSamaritanSeptemberSundaneseWednesday[:^wo
12:08:54.679 [err] roc_lib: ccc
12:08:54.679 [err] roc_lib: woooooooooooo5
=== RUN   Test_NewAddress/incorrect_ip_causes_negative_error
12:08:54.679 [err] roc_lib: woooooooooooo
12:08:54.679 [err] roc_lib: aaa --invalid ip stringlocked m0 woke upmalformed profilemark - bad statusmarkBits overflowmissing closing )missing closing ]no data availablenotetsleepg on g0permission deniedreflect.Value.Intreflect.Value.Lenreflect: New(nil)reflect: call of runtime: n
12:08:54.679 [err] roc_lib: ccc
12:08:54.679 [err] roc_lib: woooooooooooo5
--- FAIL: Test_NewAddress (0.00s)
    --- FAIL: Test_NewAddress/correct_ip_is_parsed_with_no_error (0.00s)
        roc-wrapper_test.go:27: WantErr: false, got One or more arguments are invalid
    --- PASS: Test_NewAddress/incorrect_ip_causes_negative_error (0.00s)
FAIL
FAIL	_/home/victor/dev/roc-project/roc-go/roc	0.003s
```

So instead of `127.0.0.1` roc_address_init() receives `127.0.0.1244140625: inuse: : status=Bassa_VahBhaiksukiBigEndianCuneiformDiacriticGC (idle)Hex_DigitInheritedInstMatchInstRune1InterfaceKhudawadiLINUX_2.6MalayalamMongolianNabataeanPalmyreneParseBoolParseUintSamaritanSeptemberSundaneseWednesday[:^wo...`

Let's see the generated code:

```
func addressInit(address *Address, family Family, ip string, port int32) int32 {
	...
	cip, _ := unpackPCharString(ip)
	...
}

func unpackPCharString(str string) (*C.char, *cgoAllocMap) {
	h := (*stringHeader)(unsafe.Pointer(&str))
	return (*C.char)(h.Data), cgoAllocsUnknown
}
```

So it just casts the Go string to `char*`. roc_address_init() expects a zero-terminated string (obviously), but IIRC Go strings are not zero terminated (it would be logical since we can slice them).

So we either need to make a zero-terminated string by hand or somehow tell the generator to make it for us, if it has such a feature.

File: roc/roc-wrapper.go [hunk 1]
```diff
@@ -2,18 +2,29 @@ package roc
...
-	return fmt.Errorf("%s: %d", errstr, err)
-}
+var (
+	// InvalidArguments indicates that one or more arguments passed to the function
+	// are invalid
+	InvalidArguments = fmt.Errorf("One or more arguments are invalid")
 
-// Init parses the `ip` and `port` and initializes the Address object
+	// UnknownError indicates that some not documented error appeared
+	UnknownError = fmt.Errorf("Unknown error")
```
[the remark this answers is not in the document]
Comment (Asalle) [reply roc/roc-wrapper.go]: done

File: roc/roc-wrapper.go [hunk 2]
```diff
@@ -38,48 +49,95 @@ func OpenContext(config *ContextConfig) *Context {
...
-	return convertErr(contextClose(c), "Error when closing context")
+	errCode := contextClose(c)
+	if errCode == 0 {
+		return nil
+	}
+	if errCode < 0 {
+		return InvalidArguments
+	}
+
+	return fmt.Errorf("Cannot close: there are objects attached to the context")
```
[the remark this answers is not in the document]
Comment (Asalle) [reply roc/roc-wrapper.go]: fixed

**Comment by Asalle** [comment]
holy shit! it actually does something!!

**Comment by Asalle** [comment]
I mean the go bindings :D

**Comment by Asalle** [comment]
I think we can modify the generator to use safe strings: https://github.com/xlab/c-for-go/issues/19
but now that I've modified the generated code all my edits will go to waste! D:
it's fine :disappointed:

**Comment by Asalle** [comment]
After re-gen all generated functions became exported again... I'll leave it like that right now until the bug in c-for-go will be fixed, because manually changing them after each re-gen is a terrible option...
I mean, we could add a `sed` command in makefile, but it's sad

**Comment by Asalle** [comment]
actually, scratch that, the bug has nothing to do with the functions being exported. I'll try to dig more into c-for-go or contact them.

**Comment by Asalle** [comment]
https://github.com/xlab/c-for-go/issues/83

File: roc/roc-wrapper.go [hunk 3]
```diff
@@ -1,19 +1,30 @@
...
-	}
-	return fmt.Errorf("%s: %d", errstr, err)
-}
+var (
+	// ErrInvalidArguments indicates that one or more arguments passed to the function
+	// are invalid
+	ErrInvalidArguments = errors.New("One or more arguments are invalid")
+
+	// ErrInvalidApi should never happen and indicates that the API don't follow the declared contract
+	ErrInvalidApi = errors.New("Unknown error")
```
Comment (gavv) [inline roc/roc-wrapper.go]: I guess we should change error text ("Unknown error") as well.

File: roc/roc-wrapper.go [hunk 4]
```diff
@@ -38,48 +49,94 @@ func OpenContext(config *ContextConfig) *Context {
...
 func (c *Context) Close() error {
-	return convertErr(contextClose(c), "Error when closing context")
+	errCode := contextClose(c)
+	if errCode == 0 {
+		return nil
+	}
+	if errCode < 0 {
+		return errors.New("Arguments are invalid or or there are objects attached to the context")
+	}
+	return ErrInvalidApi
```
Comment (gavv) [inline roc/roc-wrapper.go]: Nitpick: can we use the same layout in all wrappers? Either:

```
	if errCode == 0 {
		...
	}
	if errCode < 0 {
		...
	}
	return ...
```

or:

```
	if errCode == 0 {
		...
	}
	if errCode > 0 {
		...
	}
	return ...
```

**Comment by gavv** [comment]
I've added travis build to the repo. Please rebase on fresh master and it should be enabled for this PR as well.

File: roc/roc-wrapper.go (hunk 4 above)
Comment (Asalle) [reply roc/roc-wrapper.go]: done

File: roc/roc-wrapper.go (hunk 3 above)
Comment (Asalle) [reply roc/roc-wrapper.go]: done

**Comment by Asalle** [comment]
It seems that for now, we have no way of unexporting the generated functions. Let's keep it exported and fix the bug in c-for-go or hopefully somebody else will fix it first, what do you think?

**Comment by Asalle** [comment]
sorry, accidental close

**Comment by Asalle** [comment]
I fixed your comments in separate commits so it'll be easier to review. I'll fix them up if you accept this pr before merge

**Comment by gavv** [comment]
> It seems that for now, we have no way of unexporting the generated functions. Let's keep it exported and fix the bug in c-for-go or hopefully somebody else will fix it first, what do you think?

I think I've fixed that, by playing with roc.yml. I've pushed the new roc.yml to master. Now AddressSize is hidden and all functions and Frame type are unexported.

Sorry for conflicts, you'll have to rebase.

> I fixed your comments in separate commits so it'll be easier to review. I'll fix them up if you accept this pr before merge

Github allows to see what's changed from last review, so there is no need for it (at least I'm OK without it). So you can just rebase & squash.

**Comment by Asalle** [comment]
Some generated functions are exported, is that intentional?

**Comment by Asalle** [comment]
Also, this line indicates, that a `roc` folder must be in the folder with roc.yml?
`Includes: ["roc/receiver.h", "roc/sender.h", "roc/log.h"]`

btw, is master failing ci?

**Comment by gavv** [comment]
> Some generated functions are exported, is that intentional?

All functions in roc.go and now unexported. All helpers in cgo_helpers.go (like Free, PassRef, ...) are still exported. I think we can deal with them separately.

> Also, this line indicates, that a roc folder must be in the folder with roc.yml?

No. `roc/receiver.h` will match `/usr/include/roc/receiver.h` (or whenever roc is installed. It just should be in search path).

AFAIK these lines are not for c-for-go, they are just passed to cgo.

> btw, is master failing ci?

Build and test pass, coveralls complains that the coverage was decreased. So that's not a real issue.

**Comment by Asalle** [comment]
ah, my mistake, I rebased wrong. fixed now.

**Comment by gavv** [comment]
LGTM!

**Pull request merged**